glom/core.py

```diff
@@ -1334,32 +1334,62 @@ def __repr__(self):
 
     def __getstate__(self):
         t_path = _T_PATHS[self]
-        return tuple(('T' if t_path[0] is T else 'S',) + t_path[1:])
+        return tuple(({T: 'T', S: 'S', A: 'A'}[t_path[0]],) + t_path[1:])
 
     def __setstate__(self, state):
-        _T_PATHS[self] = (T if state[0] == 'T' else S,) + state[1:]
+        _T_PATHS[self] = ({'T': T, 'S': S, 'A': A}[state[0]],) + state[1:]
 
 
 _T_PATHS = weakref.WeakKeyDictionary()
 
 
 def _t_child(parent, operation, arg):
     t = TType()
-    _T_PATHS[t] = _T_PATHS[parent] + (operation, arg)
+    base = _T_PATHS[parent]
+    if base[0] is A and operation not in ('.', '[', 'P'):
+        # whitelist rather than blacklist assignment friendly operations
+        # TODO: error type?
+        raise BadSpec("operation not allowed on A assignment path")
+    _T_PATHS[t] = base + (operation, arg)
     return t
 
 
+def _s_first_magic(scope, key, _t):
+    """
+    enable S.a to do S['a'] or S['a'].val as a special
+    case for accessing user defined string variables
+    """
+    err = None
+    try:
+        cur = scope[key]
+    except (KeyError, IndexError) as e:
```

Comment:
wait, how is it possible that IndexError get raised here? Is it possible for scope to be something other than a dict?

Reply:
(removed it, tests all passing of course)

Reply:
(kurt lgtm'd offline)